### DISCUSSION TITLE
ScopedModuleNameOverride

### DISCUSSION THREAD

**Discussion by tdegeus** [pybind/pybind11#3748] [Ideas]
I have a set of functions written in C++ that I compile in a module `_foo`.
My python module simply imports all functions from it. I.e. `foo/__init__.py` reads
```python
from _foo import *
```

This was giving issues with my documentation. And I got the suggestion https://github.com/sphinx-doc/sphinx/issues/10199 to use:
```cpp
/**
Overrides the `__name__` of a module.
Classes defined by pybind11 use the `__name__` of the module as of the time they are defined,
which affects the `__repr__` of the class type objects.
*/
class ScopedModuleNameOverride {
public:
    explicit ScopedModuleNameOverride(py::module m, std::string name) : module_(std::move(m))
    {
        original_name_ = module_.attr("__name__");
        module_.attr("__name__") = name;
    }
    ~ScopedModuleNameOverride()
    {
        module_.attr("__name__") = original_name_;
    }

private:
    py::module module_;
    py::object original_name_;
};

PYBIND11_MODULE(_foo, mod)
{
    ScopedModuleNameOverride name_override(m, "foo");
}
```
See https://github.com/google/tensorstore/blob/94be4f2e8715511bb60fc0a0eaf07335881673b3/python/tensorstore/tensorstore.cc#L75

It would be great if this could be incorporated as option in pybind11 !

**Comment by rwgk** [comment]
At first sight I thought this looks interesting, but at second thought, in-place modifications like this usually come with pitfalls & surprises. (BTW: I'd definitely delete the copy & move constructor & operator in the tensorstore implementation; it'll be surprising and bug-prone if the object is accidentally copied or moved.)

I think it would be safer and more obvious to create a `module_alias`, a new object, which has all the attributes we need,  exactly the same is `m`, except `__name__`.

```
auto ma = m.make_alias("foo");
```
or
```
auto ma = m.make_alias(py::module_::strip_leading_underscore_from_name);
```

Then use `ma` instead of `m` for the `.def`s etc. That will be completely obvious, and much less surprising to someone glancing through the code while debugging, for example (i.e. with the mind somewhere else, which is >90% how I usually read code).

**Reply by tdegeus** [reply]
`make_alias` seems like a great solution! And the helper too! I would open a PR if I would know how/where to implement this (which I sadly dont‘t)

**Comment by rwgk** [comment]
This is an OSS project. Try!

**Comment by tdegeus** [comment]
@rwgk Am I on the right track if I put:
```cpp
    struct strip_leading_underscore_from_name;

    static module_ make_alias(const char *name) {
        module_ ret = *this;
        ret.attr("__name__") = this.attr(name);
        return ret;
    }

    static module_ make_alias(strip_leading_underscore_from_name) {
        auto name = ??
        return this->make_alias(name);
    }
```
as members of `module_`?

**Comment by tdegeus** [comment]
Here is my (potentially naive ;)) try : https://github.com/pybind/pybind11/pull/3761
